deepali/core/image.py

```diff
@@ -1163,6 +1164,75 @@ def grid_sample_mask(
     )
 
 
+def rand_sample(
+    data: Union[Tensor, Sequence[Tensor]],
+    num_samples: int,
+    mask: Optional[Tensor] = None,
+    replacement: bool = False,
+    generator: Optional[Generator] = None,
+) -> Union[Tensor, Sequence[Tensor]]:
+    r"""Random sampling of voxels within an image.
+
+    Args:
+        data: One or more image batch tensors with shape ``(N, C, ... X)`` to sample
+            from at the same random image grid positions, i.e., voxel indices. Note that
+            all input tensors must have the same number and size of spatial dimensions.
+        num_samples: Number of spatial samples to draw.
+        mask: Optional mask to use for spatially weighted sampling.
+        replacement: Whether to sample with or without replacement.
+        generator: Random number generator to use.
+
+    Returns:
+        Tensor of shape ``(N, C, num_samples)`` with input ``data`` values at randomly
+        sampled spatial grid points. When ``data`` is a sequence of tensors, a list
+        of tensors with order matching the input data is returned.
+
+    """
+    input: Sequence[Tensor]
+    if isinstance(data, Tensor):
+        input = [data]
+    elif isinstance(data, Sequence) and all(isinstance(x, Tensor) for x in data):
+        input = data
+    else:
+        raise TypeError("rand_sample() 'data' must be Tensor or Sequence[Tensor]")
+    if not input:
+        return []
+    if any(x.ndim < 3 for x in input):
+        raise ValueError("rand_sample() 'data' must be one or more tensors of shape (N, C, ..., X)")
+    shape = input[0].shape
+    device = input[0].device
+    if any(x.shape[2:] != shape[2:] for x in input):
+        raise ValueError("rand_sample() 'data' tensors must have identical spatial shape")
+    numel = shape[2:].numel()
+    if not replacement and num_samples > numel:
+        raise ValueError("rand_sample() 'num_samples' is greater than number of spatial points")
+    input = [x.flatten(2) for x in input]
+    if not replacement and mask is None:
+        perm = torch.empty(numel, dtype=torch.int64, device=device)
+        index = torch.empty((shape[0], num_samples), dtype=torch.int64, device=device)
+        for row in index:
+            torch.randperm(numel, generator=generator, out=perm, device=device)
```

Comment:
Considering timings reported at https://discuss.pytorch.org/t/torch-equivalent-of-numpy-random-choice/16146/19, maybe best to perform `randperm` on CPU and/or use one of the faster alternative approaches noted.